### ISSUE TITLE
Question about partitioning options

### ISSUE THREAD

**Issue by pferrel** [scopt/scopt#39]
I need to repeat the same options in several CLI drivers, Specifically they contain options for how to parse a text delimited (or later other text formatted files like JSON). The portion of options dealing with the schema of the input and output files will be repeated in each of several CLIs.

Also the companion "Config" case class  (called Options in this example) that contains defaults will be partially repeated.

This makes for a problem since the repeated code may change and is itself not completely trivial with consistency checks. 

Can you suggest a good way to partitiion options into groupings and then pull in one class or function that parses that group of options and/or Config values

An example of the not yet complete CLI help output is [here](https://github.com/pferrel/harness/wiki#example-1-itemsimilairty)

**Comment by pferrel** [comment]
Each of the OptionParser methods of interest return an OptionDef, seems like I could put some helper functions in the package object to process certain groups of options and reuse them in other drivers?

**Comment by metasim** [comment]
I'm also interested in this question.

**Comment by metasim** [comment]
In case there's interest, the typeclass pattern can be used to modularize configuration without modifying the existing scopt implementation.:

``` scala
/**
 * JDBC configuration container case class aimed for use with `scopt`
 */
case class JdbcConfig(
    driver: String = "org.postgresql.Driver",
    conn: String = "jdbc:postgresql:mydb",
    user: String = "postgres",
    pass: String = "")

/**
 * Typeclass declaration for defining required support of updating nested
 * configuration objects.
 * For example, if you have an outer config class such as this:
 *
 * {{{
 * case class Config(someThing: String, jdbc: JdbcConfig = JdbcConfig())
 * }}}
 *
 * Create an implicit definition of `JdbcConfigOwner[Config]` so `JdbcConfig.inject`
 * can know how to update the configuration:
 *
 * {{{
 * implicit object IHazJdbc extends JdbcConfigOwner[Config] {
 *      override def extract(c: Config): JdbcConfig = c.jdbc
 *      override def update(c: Config, jdbc: JdbcConfig): Config = c.copy(jdbc = jdbc)
 * }
 * }}}
 *
 * Then just call `JdbcConfig.inject`:
 *
 * {{{
 * val argParser = new scopt.OptionParser[Config]("FooBar") {
 *     help("help").text("prints this usage text")
 *     // etc...
 *     ...
 *     JdbcConfig.inject(this)
 * }
 * }}}
 */
trait JdbcConfigOwner[C] {
    def extract(c: C): JdbcConfig
    def update(c: C, conf: JdbcConfig): C
}

/** Companion object for `JdbcConfig` configuration command line parser support. */
object JdbcConfig {
    def inject[T : JdbcConfigOwner](parser: scopt.OptionParser[T]) {
        // Import parameter so parser config looks like a top-level implementation
        import parser._

        // Inject typeclass magic
        val jcoTC = implicitly[JdbcConfigOwner[T]]
        import jcoTC._

        opt[String]("jdbcConnection").optional().
            text("JDBC connection URL").
            action((conn, conf) ⇒ update(conf, extract(conf).copy(conn = conn)))

        opt[String]("jdbcDriver").optional().
            text("JDBC driver class name").
            action((driver, conf) ⇒ update(conf, extract(conf).copy(driver = driver)))

        opt[String]("dbUser").optional().
            text("Database user name").
            action((user, conf) ⇒ update(conf, extract(conf).copy(user = user)))

        opt[String]("pass").optional().
            text("Database user password").
            action((pass, conf) ⇒ update(conf, extract(conf).copy(pass = pass)))
    }

    /** Self-referential type class for drivers that only have `JdbcConfig`. */
    implicit object JdbcConfigIsOwner extends JdbcConfigOwner[JdbcConfig] {
        override def extract(conf: JdbcConfig): JdbcConfig = conf
        override def update(conf: JdbcConfig, jdbc: JdbcConfig): JdbcConfig = jdbc
    }
}
```

**Comment by eed3si9n** [comment]
Maybe I should add a feature to append options from another parser. That should take care of the composition of the parser. The second piece would the composition of config case classes. I'm not familiar with it, but maybe Shapeless's `LabelledGeneric` could be used to implement `updateWith`?

``` scala
def parser1[A: UpdateWith]: scopt.OptionParser = new scopt.OptionParser[A]("scopt") {
  opt[String]("jdbcConnection") optional() text("JDBC connection URL") action { (x, c) =>
    updateWith(c, 'jdbcConnection, x)
  }
}
```

**Comment by metasim** [comment]
Interesting idea. I'm not familiar with `LabelledGeneric` either, but the resulting DSL is nice. I like using `Symbol`s that way. If you have to rely on a third-party library, that's not a bad one, but I do like the fact that currently _scopt_ is effectively dependency free.

I can't help but to wonder if generalizing the `UpdateWith[A]` concept further might allow for a set of nice defaults whereby a lot of the user's case class `copy` boilerplate could be eliminated somehow in standard usage, as well as facilitate this modular, nested case class capability.

**Comment by pferrel** [comment]
What I did is pretty simplistic. I keep my options in several Maps. Each are initialized with default values for groups of common options. In an extending MyOptionParser I put a bunch of helper methods that parse the common option groups. This works pretty well and is implemented in the immutable idiom. The options being used in any particular case can be combined with the Map `++` append in the driver. Maps that contain the same mappings are resolved by append order so if you have a different default for the driver it can override the boilerplate default.

```
object ItemSimilarityDriver extends MyDriver {
  // define only the options specific to ItemSimilarity
  private final val ItemSimilarityOptions = HashMap[String, Any](
    "maxPrefs" -> 500,
    "maxSimilaritiesPerItem" -> 100,
    "appName" -> "DriverSpecificName")// this overrides the default option

  private var options: Map[String, Any] = null

  /**
   * @param args  Command line args, if empty a help message is printed.
   */
  override def main(args: Array[String]): Unit = {
    // build options from some stardard CLI param groups
    // Note: always put the driver specific options at the last so they can override less specific defaults!
    options = MyOptionParser.GenericOptions ++ MyOptionParser.SparkOptions ++  ItemSimilarityOptions

    val parser = new MyOptionParser(programName = "spark-mydriver") {
      head("spark-mydriver", "Some Job")

      //Input output options, non-driver specific
      // this is in MyOptionParser since it handles standards driver options, saves a lot of boilerplate
      parseIOOptions

```

Then MyOptionsParser has all the default Maps created and contains parsing for boilerplate parsing and validations. `parseIOOptions` is a method that handles parsing, usage messages and validations for that option group.

One problem with this is that every time you read from an option it must be cast.`MyOptionParser.GenericOptions` is a `HashMap[String, Any]` so any access must be of the form `options("input").asInstanceOf[String]` which can get rather verbose.

I considered several ways to use case classes of case classes to avoid the casting but you have the problem that the option container's type must be known at the time the option parser is created, so a Map solved that. I'm pretty satisfied with this but am not a Scala wizard so there may be a better way.

If @eed3si9n suggestion only composes the options containers it should avoid the casting? Not familiar with `LabelledGeneric`.

**Comment by aappddeevv** [comment]
This is a bit old item, but I needed to do this recently. Since the dsl (opt, arg, etc.) mutates internally and "add" order is significant, the best option I had was to simply write a function that takes a ConfigParser and calls opt/arg/... where needed.

```scala
def generalArgs(op: scopt.OptionParser[MyConfig]): Unit = {
   import op._
  opt[String]("debug").text("..").hidden().action(...)
}
```
You can create a small builder pattern if you want to make it seem immutable. You could compose further with more cleverness. Since the config object has to be known to be used in the .action clause, you have to work a bit harder at it.

**Comment by eed3si9n** [comment]
Released scopt 4.0.0-RC2 with more composable OParser. See http://eed3si9n.com/scopt4 for details.
